### PR TITLE
Check threshold of role type

### PR DESCRIPTION
**Pull request by july2993**
cc #224 #229 

check all signatures must be correct and has at lest threshold number according to role type.
move how about get keys outside local manifest

### CODE REVIEW

File: pkg/repository/v1manifest/manifest.go [hunk 1]
```diff
@@ -293,9 +297,54 @@ func ReadManifest(input io.Reader, role ValidManifest, keys crypto.KeyStore) (*M
 		return nil, errors.New("no signatures supplied in manifest")
 	}
 
-	err = m.VerifySignature(uint(len(m.Signatures)), keys)
-	if err != nil {
-		return nil, err
+	if role.Base().Ty != ManifestTypeRoot {
+		if root != nil {
```
[the remark this answers is not in the document]
Comment (july2993) [reply pkg/repository/v1manifest/manifest.go:301]: Yes, the `FsManifests.LoadManifest` will call this to load manifest too, and the previous just pass a nil KeyStore will not check any signature too actually.
It seems no need the check signature when loading the local manifest? Or we need to pass the `root` to `FsManifests.LoadManifest` too.

File: pkg/repository/v1manifest/manifest.go [hunk 2]
```diff
@@ -293,9 +297,54 @@ func ReadManifest(input io.Reader, role ValidManifest, keys crypto.KeyStore) (*M
...
+		if err != nil {
+			return nil, errors.AddStack(err)
+		}
+		threshold := newR.Roles[ManifestTypeRoot].Threshold
+
+		if !checkHasThresholdKeys(threshold, m.Signatures, keys) {
+			return nil, errors.Errorf("no enough %d signature: %v, keys: %v", threshold, m.Signatures, keys)
+		}
+
+		if root != nil {
```
[the remark this answers is not in the document]
Comment (july2993) [reply pkg/repository/v1manifest/manifest.go:325]: currently, only reading at local will pass nil.
will not pass nil when reading from mirror

File: pkg/repository/v1manifest/manifest.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (nrc) [reply pkg/repository/v1manifest/manifest.go:301]: We should check the signature when loading locally, unless it is very expensive. I think it is fine for the FsManifest to keep a reference to a key store (or root manifest)

File: pkg/repository/v1manifest/manifest.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (nrc) [reply pkg/repository/v1manifest/manifest.go:325]: Although the code is correct, I think it is too easy for future developers to make a mistake and that could easily cause a critical vulnerability.

**Comment by july2993** [comment]
> Looks good, a few comments inline. I was thinking that the key store would hold all the keys that are being used, i.e., from all root manifests and the index file and then logic for finding keys and ensuring we are using the right ones can live there. I'm not sure if that actually makes it nicer though.

For Component, we are not checking the signature now, I will fix this as flow up, maybe fix not checking the local manifest too.

**Pull request merged**